CppCustomVisualizer2/dll/RootVisualizer.cpp

```diff
@@ -0,0 +1,320 @@
+    return S_OK;
+}
+
+//static 
+HRESULT CRootVisualizer::CreateEvaluationResult(_In_ DkmVisualizedExpression* pVisualizedExpression, _Deref_out_ DkmEvaluationResult** ppResultObject)
+{
+    HRESULT hr = S_OK;
+
+    CComObject<CRootVisualizer>* pRootVisualizer;
+    CComObject<CRootVisualizer>::CreateInstance(&pRootVisualizer);
+    pRootVisualizer->Initialize(pVisualizedExpression);
+    pVisualizedExpression->SetDataItem(DkmDataCreationDisposition::CreateNew, pRootVisualizer);
+
+    CComPtr<DkmRootVisualizedExpression> pRootVisualizedExpression = DkmRootVisualizedExpression::TryCast(pVisualizedExpression);
+    if (pRootVisualizedExpression == nullptr)
+    {
+        // This sample doesn't provide child evaluation results, so only root expressions are expected
+        return E_NOTIMPL;
+    }
+    CComPtr<DkmString> pName = pRootVisualizedExpression->Name();
+    CComPtr<DkmString> pFullName = pRootVisualizedExpression->FullName();
+    DkmRootVisualizedExpressionFlags_t flags = pRootVisualizedExpression->Flags();
+
+    hr = pRootVisualizer->CreateEvaluationResult(
+        pName,
+        pFullName,
+        pRootVisualizedExpression->Type(),
+        flags,
+        nullptr,
+        pVisualizedExpression->InspectionContext(),
+        ppResultObject
+    );
+
+    return hr;
+}
+
+HRESULT CRootVisualizer::CreateEvaluationResult(
+    _In_ DkmString* pName,
+    _In_ DkmString* pFullName,
+    _In_ DkmString* pType,
+    _In_ DkmRootVisualizedExpressionFlags_t flags,
+    _In_opt_ DkmVisualizedExpression* pParent,
+    _In_ DkmInspectionContext* pInspectionContext,
+    _Deref_out_ DkmEvaluationResult** ppResultObject
+)
+{
+    HRESULT hr = S_OK;
+
+    CComPtr<DkmPointerValueHome> pPointerValueHome = DkmPointerValueHome::TryCast(m_pVisualizedExpression->ValueHome());
+    if (pPointerValueHome == nullptr)
+    {
+        // This sample only handles visualizing in-memory Sample structures
+        return E_NOTIMPL;
+    }
+
+    // Create method for DkmDataAddress takes a runtime instance.
+    CComPtr<DkmDataAddress> pAddress;
+    hr = DkmDataAddress::Create(m_pVisualizedExpression->RuntimeInstance(), pPointerValueHome->Address(), NULL, &pAddress);
+    if (FAILED(hr))
+    {
+        return hr;
+    }
+
+    // Read the Sample value from the target process
+    // This is NOT a deep read, so we basically can only get the size of the vectors by reading the memory directly.
+    DkmProcess* pTargetProcess = pInspectionContext->RuntimeInstance()->Process();
+    BYTE sampleRaw[sizeof(Sample)];
+    hr = pTargetProcess->ReadMemory(pPointerValueHome->Address(), DkmReadMemoryFlags::None, sampleRaw, sizeof(Sample), nullptr);
+    if (FAILED(hr))
+    {
+        // If the bytes of the value cannot be read from the target process, just fall back to the default visualization
+        return E_NOTIMPL;
+    }
+    Sample* pSampleValue = (Sample*)(void*)sampleRaw;
+
+    if (pSampleValue->a.size() != pSampleValue->b.size())
```

Comment:
>if (pSampleValue->a.size() != pSampleValue->b.size()) [](http://example.com/codeflow?start=4&length=53)

This is dangerous since:
1. If you debug a different bitness, this layout will be wrong
2. Even in the same bitness, the STL could change the layout of `vector`.

I think you should ask the C++ EE for the size of it instead.